### PR TITLE
New data channel based on QNX Messaging

### PR DESCRIPTION
**Pull request by svstanko**
- Add a new DLT-IPC "QNX_MESSAGE" based on the QNX messaging IPC
in addition to the two existing DLT-IPC: FIFO and UNIX_SOCKET.

The problem:
Message passing is the primary form of IPC in the QNX Neutrino
RTOS. POSIX IPC (TCP/IP sockets and pipes) are built over QNX
native message passing and sometimes it can cause 2 copies of
data, more context switches.

The solution:
Using of Resource Manager helps to simplify client-server
communication and switch DLT to QNX Neutrino Messaging instead
Unix sockets with minimum efforts without changing in DLT internal
logic.

The conclusions:
The main benefit of replacing Unix sockets with QNX Neutrino
Messaging is significant unloading of the networking manager
(io-pkt-v6-hc):
- CPU Usage ~ 9 times better;
- Inter CPU Communication (Cross CPU Messages) ~68 times better;
- CPU Migration ~ 2 times better.

DLT daemon utilizes more CPU recourses (instead of the network
driver), but this has not degraded overall performance.

Noticeable improvement in CPU usage has been achieved.

### CODE REVIEW

File: CMakeLists.txt
Comment (ssugiura) [inline CMakeLists.txt:70]: If the user set to QNX_MESSAGE, we would verify first that it is compiled under QNX, otherwise raise an error and return already. You could do something similar to line 161. Then you don't need to consider QNX and non-QNX inside headers and codes.

File: include/dlt/dlt_common.h
Comment (ssugiura) [inline include/dlt/dlt_common.h:115]: As commented above, we don't require this ifdef, no error message at line 120 either. The same for other places.

File: src/daemon/dlt-daemon.c
Comment (ssugiura) [inline src/daemon/dlt-daemon.c:82]: We need this only for DLT_DAEMON_USE_QNX_MESSAGE_IPC, I guess

File: src/daemon/dlt_daemon_common.c
Comment (ssugiura) [inline src/daemon/dlt_daemon_common.c:99]: Do we need this even if QNX_MESSAGE_IPC is not set?

File: src/lib/dlt_user.c
Comment (ssugiura) [inline src/lib/dlt_user.c:298]: Correct indent

File: src/lib/dlt_user.c
Comment (ssugiura) [inline src/lib/dlt_user.c:272]: Correct indent

**Review by ssugiura — CHANGES_REQUESTED** [review]
Overall implementation looks fine with me.
I also would like to request you to document overall design and sequence of QNX messaging.
Could you prepare a new document with e.g. `doc/dlt_ipc_qnx_message.md` and describe this?
Afterwards I'd like to add documents for other IPCs (FIFO and unix socket).
This would definitely help new comers to decide which IPC to be used.
Thanks in advance!

**Comment by minminlittleshrimp** [comment]
Hello @svstanko 

Do you still have any interest in this topic?
If yes could you please let me know if you could resolve the conflict so we could continue reviewing for you?

Thank you

**Comment by michael-methner** [comment]
Closing pull request as there seems to be no interest of the author anymore.
Please reopen if you want to rework it.

**Comment by alexmohr** [comment]
@marcin-sochacki As you're employed by BlackBerry it would be awesome if you pickup the slack here. Getting message passing to work fully reduces the load in dlt-daemon tremendously. 
We did some work internally to improve on this MR and it showed reductions in CPU usage of up to 30%. Unfortunately we never got it fully running. 

Imho this would be huge win in the context of "QNX everywhere" :wink:

**Comment by marcin-sochacki** [comment]
@alexmohr @minminlittleshrimp @michael-methner 
This is indeed a great performance improvement!
I would like to understand if dlt-daemon is going to be used in the future automotive projects (and if that optimization would help the customers) or is t just maintained for existing projects (bug fixes, etc.).
This "Pull Request" was created in 2021 and the code base changed a lot since then, so I expect that rebasing may not be a trivial task.
Anyway I could look at it, but the precondition is another pull request with build system changes for QNX:
https://github.com/COVESA/dlt-daemon/pull/609

**Comment by alexmohr** [comment]
Obviously I can't speak too much about in which direction development is going for us, but for the foreseeable future dlt-daemon is likely to be a central component in the logging architecture. 
Even for current generations getting this improvement, would be a huge benefit for us and probably everyone else who is using dlt-daemon on QNX.

**Comment by michael-methner** [comment]
Hello @marcin-sochacki ,
i cannot disclose any details on our current projects but I would second @alexmohr statements. Having this performance improvements on QNX would be great as DLT on QNX is often used as central logging component in multi-os setups.

**Comment by svstanko** [comment]
Hello @michael-methner,
Sorry for delay with the answer.
That solution was abandoned after deeper investigation. It passed all our tests but failed with more complicated scenarios of other components.
Tests failed due to unexpected stopping of dlt-deamon in some test cases with stopping several components that uses dlt. (Stopping only component didn't affect dlt in such way).
The root cause of the issue:
When the daemon receives a request to set (change) the log level, it tries to send this request to all clients, and when trying to send such a request to clients that have already disconnected, an error occurs, according to which the daemon exits its event processing loop.
The solution hasn't been found.

**Comment by marcin-sochacki** [comment]
@alexmohr 
QNX would be happy to help.
So far I re-based the commit onto recent software baseline, applied the changes from the code review and fixed a dependency issue, which didn't allow to build unit tests for QNX:
https://github.com/marcin-sochacki/dlt-daemon/tree/resource_manager
I run unit tests under QNX in QEMU and some of them fail.
Since this is my background task, it may take some time, but I'm going to investigate the issues and try to fix them. Until now, you may try it out from my fork.

**Comment by alexmohr** [comment]
@marcin-sochacki That is great news, thank you! 

Can you post which tests are failing? We are running unit tests on qnx internally as well but did some patches to the tests to fix some issues. So maybe we fixed these issues already and can upstream the required changes.

We never bothered to upstream so far these because the changes can't be validated in the github pipeline anyways.

**Comment by marcin-sochacki** [comment]
Here are the test reports: 
[test_reports.zip](https://github.com/COVESA/dlt-daemon/files/15309115/test_reports.zip)

**Comment by alexmohr** [comment]
We actually had to implement quite some changes to get these to run. 
https://github.com/mercedes-benz/dlt-daemon/commit/f5de8907ef8080ec601185386fb031efdcc72a70

**Comment by marcin-sochacki** [comment]
> Our issue with this was that clients running with a higher priotity than dlt-daemon get degraded to the dlt-daemon priotity while sending messages. Through the priotity inversion cricitical processes don't get CPU time anymore (or too late) resulting in a watchdog kill.

QNX uses priority inheritance to actually prevent problems of high priority clients blocked on servers servicing low priority clients. QNX boost the server's priority to be the highest of all blocked clients:
https://www.qnx.com/developers/docs/7.1/#com.qnx.doc.neutrino.getting_started/topic/s1_msg_prio_interitance.html
However there are some exceptions, for example if threads use semaphores (instead of mutexes) for synchronization.
I will take a look at the above changes. As mentioned before, I only re-based the commit onto recent software baseline and fixed obvious compilation issues, but didn't look at the content so far.

**Comment by marcin-sochacki** [comment]
I compared:
https://github.com/marcin-sochacki/dlt-daemon/tree/resource_manager
with
https://github.com/mercedes-benz/dlt-daemon/tree/message-passing
and it looks that your branch contains also other changes, making it more difficult to have a common codebase.

My quick findings:
1) dlt_receiver_receive()
function may not always reply to client's MsgSend(), keeping them blocked.
There are several if else statements there and MsgReply() is called only for one condition.
In other cases, the client should also get an answer - MsgError() should be called to unblock the client.
2) The Resource Manager shall implement close_dup handler to clean-up client-oriented resources,
otherwise the Resource Manager does not know if a client died or simply closed the connection.
3) And as already mentioned: Semaphores don't affect a thread's effective priority; if you need priority inheritance, use a mutex.

I hope that helps.

**Pull request closed without merging**
